### ISSUE TITLE
ANUGA GPU support

### ISSUE THREAD

**Issue by samcom12** [anuga-community/anuga_core#35]
Hi All,

With recent developments like `OpenMP offload` support or `openacc pragma`, Can we port ANUGA to use GPU and accelerate simulations?

Cheers,
Samir

**Comment by stoiver** [comment]
@samcom12 good idea. I have been thinking about this as well. The first step I think is to convert the inner timestepping procedures to cython C and then invoking openmp in the various expensive procedures. Hopefully we could then move over to offloading to the GPU.

**Comment by samcom12** [comment]
Thanks @stoiver .
I just tried profiling for offloading benefits with sample `rectangular `testcase. The results are promising showing almost 4X speedup which can further increase with real datasets we are using.

You can refer the attached report.
[offload_model.tar.gz](https://github.com/anuga-community/anuga_core/files/10695646/offload_model.tar.gz)

**Comment by stoiver** [comment]
@samcom12, the results look promising. It is good that the offload advisor just picks up 2 procedures, _compute_fluxes_central and _extrapolate_second_order_edge_sw to look at. I think it also suggests that we should look at just using standard openmp on those two procedures. Does intel have an advisor for studying likely improvement using openmp?

**Comment by samcom12** [comment]
Hello @stoiver 

There is a tool which converts serial C,C++, and Fortran code to OpenMP based code.

http://rosecompiler.org/

And, with Intel VTune we can do threading analysis to see how effectively threads are being used in the code.

https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/top/analyze-performance/parallelism-analysis-group/threading-analysis.html

**Comment by JorgeG94** [comment]
Hi @samcom12 we have picked openmp offloading for GPU porting, currently we're waiting on Meson to do a release with a bug fix that will allow us to use the nvc compiler. The reason to not use gcc at the moment is that the nvidia tools won't work well if we don't use the nvidia compiler. 

Much of the code has been refactored and we've eliminated all but the OpenMP and the CUDA code from ANUGA - we might end up writing CUDA kernels if we find very very expensive functions, otherwise we might end up just with OpenMP.

**Comment by JorgeG94** [comment]
In theory OMP offloading should be great since it should be the same across compilers, we just depend on compiler support being good - the good thing is that it is getting way better. 

I don't really like kokkos in general but I am open to it if you give me solid examples; this is just personal preference :)
